cirq-core/cirq/devices/thermal_noise_model.py

```diff
@@ -0,0 +1,251 @@
+        """
+        qubits = set(qubit_dims)
+        rate_dict = {}
+
+        def _as_rate_dict(
+            rate_or_dict: Optional[Union[float, Dict['cirq.Qid', float]]]
+        ) -> Dict['cirq.Qid', float]:
+            # Convert float or None input into dictionary form. Make sure no
+            # qubits are missing from dictionary input.
+            if rate_or_dict is None:
+                return {qb: 0.0 for qb in qubits}
+            elif isinstance(rate_or_dict, dict):
+                out = rate_or_dict.copy()
+                for qb in qubits:
+                    if qb not in rate_or_dict:
+                        out[qb] = 0.0
+                return out
+            else:
+                return {qb: rate_or_dict for qb in qubits}
+
+        heat_rate_GHz = _as_rate_dict(heat_rate_GHz)
+        cool_rate_GHz = _as_rate_dict(cool_rate_GHz)
+        dephase_rate_GHz = _as_rate_dict(dephase_rate_GHz)
+
+        for q, dim in qubit_dims.items():
+            gamma_h = heat_rate_GHz[q]
+            gamma_c = cool_rate_GHz[q]
+            gamma_phi = dephase_rate_GHz[q]
+
+            rate_dict[q] = _decoherence_matrix(gamma_c, gamma_phi, gamma_h, dim)
+
+        _validate_rates(qubit_dims, rate_dict)
+        self.gate_durations_ns: Dict[type, float] = gate_durations_ns
+        self.rate_matrix_GHz: Dict['cirq.Qid', np.ndarray] = rate_dict
+        self.require_physical_tag: bool = require_physical_tag
+        self.skip_measurements: bool = skip_measurements
+
+    def noisy_moment(
+        self, moment: 'cirq.Moment', system_qubits: Sequence['cirq.Qid']
+    ) -> 'cirq.OP_TREE':
+        noise_ops: List['cirq.Operation'] = []
+        moment_ns: float = 0
+        for op in moment:
+            op_duration: Optional[float] = None
+            for key, duration in self.gate_durations_ns.items():
+                if not issubclass(type(op.gate), key):
+                    continue  # gate type doesn't match
+                # TODO: remove assumption of same time across qubits
+                # if len(key) > 1 and op_data[:1] != key[:1]:
+                #     continue  # qubits don't match
+                op_duration = duration
+                break
+            if op_duration is None:
+                if not isinstance(op.gate, ops.WaitGate):
+                    continue
+                # special case for wait gates if not predefined
+                op_duration = op.gate.duration.total_nanos()
+            moment_ns = max(moment_ns, op_duration)
+
+        if moment_ns == 0:
+            return [moment]
+
+        for qubit in system_qubits:
+            qubit_op = moment.operation_at(qubit)
+            if qubit_op is None:
+                continue
+            if self.skip_measurements and protocols.is_measurement(qubit_op):
+                continue
+            if self.require_physical_tag and PHYSICAL_GATE_TAG not in qubit_op.tags:
+                # Only non-virtual gates get noise applied.
+                continue
```

Comment:
Could we convert all these guards into list comprehensions ?
```python3
valid_qubit_ops = [qubit for qubit in system_qubits if moment.operation_at(qubit) is not None]
.
.


```

Reply:
Inside the loop we need the qubits, not the operations. This leads to a comprehension like:
```
affected_qubits = [
    q for q in system_qubits if not any(
        moment.operation_at(q) is None,
        self.skip_measurements and protocols.is_measurement(moment.operation_at(q)),
        self.require_physical_tag and PHYSICAL_GATE_TAG not in moment.operation_at(q).tags
    )
]
```
Setting aside the additional cost of calling `moment.operation_at(q)` multiple times, I don't think this is an improvement in readability over the current guards.